chain/db_test.go

```diff
@@ -0,0 +1,293 @@
+package chain
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+	"lukechampine.com/frand"
+)
+
+// StandardUnlockConditions returns the standard unlock conditions for a single
+// Ed25519 key.
+func StandardUnlockConditions(pub types.PublicKey) types.UnlockConditions {
+	return types.UnlockConditions{
+		PublicKeys: []types.UnlockKey{{
+			Algorithm: types.SpecifierEd25519,
+			Key:       pub[:],
+		}},
+		SignaturesRequired: 1,
+	}
+}
+
+// StandardAddress returns the standard address for an Ed25519 key.
+func StandardAddress(pub types.PublicKey) types.Address {
+	return StandardUnlockConditions(pub).UnlockHash()
+}
+
+// AppendTransactionSignature appends a TransactionSignature to txn and signs it
+// with key.
+func AppendTransactionSignature(cs consensus.State, key types.PrivateKey, parentID types.Hash256, txn *types.Transaction) {
+	sig := types.TransactionSignature{
+		ParentID:       parentID,
+		CoveredFields:  types.CoveredFields{WholeTransaction: true},
+		PublicKeyIndex: 0,
+	}
+	sigHash := key.SignHash(cs.WholeSigHash(*txn, sig.ParentID, sig.PublicKeyIndex, sig.Timelock, sig.CoveredFields.Signatures))
+	sig.Signature = sigHash[:]
+
+	txn.Signatures = append(txn.Signatures, sig)
+}
+
+// FindBlockNonce finds a block nonce meeting current target.
+func FindBlockNonce(cs consensus.State, b *types.Block) {
+	// ensure nonce meets factor requirement
+	for b.Nonce%cs.NonceFactor() != 0 {
+		b.Nonce++
+	}
+	for b.ID().CmpWork(cs.ChildTarget) < 0 {
+		b.Nonce += cs.NonceFactor()
+	}
+}
+
+func addBlock(cm *Manager, b *types.Block) error {
+	var minerAddress types.Address
+	frand.Read(minerAddress[:])
+
+	cs := cm.TipState()
+	b.MinerPayouts = []types.SiacoinOutput{{Address: minerAddress, Value: cs.BlockReward()}}
+	FindBlockNonce(cs, b)
+	return cm.AddBlocks([]types.Block{*b})
+}
+
+type diffSubscriber struct {
+	cau *consensus.BlockDiff
+	cru *consensus.BlockDiff
+}
+
+func (s *diffSubscriber) ProcessChainApplyUpdate(cau *ApplyUpdate, mayCommit bool) error {
+	if cau == nil {
```

Comment:
`cau` and `cru` will never be nil -- they are passed by pointer for efficiency (although perhaps they shouldn't be...)